keras_cv/models/object_detection/faster_rcnn.py

```diff
@@ -0,0 +1,326 @@
+
+    def build(self, input_shape):
+        if isinstance(input_shape, (dict, list, tuple)):
+            input_shape = tf.nest.flatten(input_shape)
+            input_shape = input_shape[0]
+        filters = input_shape[-1]
+        self.conv = tf.keras.layers.Conv2D(
+            filters=filters,
+            kernel_size=3,
+            strides=1,
+            padding="same",
+            activation="relu",
+            kernel_initializer="truncated_normal",
+        )
+        self.objectness_logits = tf.keras.layers.Conv2D(
+            filters=self.num_anchors * 1,
+            kernel_size=1,
+            strides=1,
+            padding="same",
+            kernel_initializer="truncated_normal",
+        )
+        self.anchor_deltas = tf.keras.layers.Conv2D(
+            filters=self.num_anchors * 4,
+            kernel_size=1,
+            strides=1,
+            padding="same",
+            kernel_initializer="truncated_normal",
+        )
+
+    def call(self, feature_map):
+        def call_single_level(f_map):
+            batch_size = f_map.get_shape().as_list()[0]
+            if batch_size is None:
+                raise ValueError("Cannot handle static shape")
+            # [BS, H, W, C]
+            t = self.conv(f_map)
+            # [BS, H, W, K]
+            rpn_scores = self.objectness_logits(t)
+            # [BS, H, W, K * 4]
+            rpn_boxes = self.anchor_deltas(t)
+            # [BS, H*W*K, 4]
+            rpn_boxes = tf.reshape(rpn_boxes, [batch_size, -1, 4])
+            # [BS, H*W*K, 1]
+            rpn_scores = tf.reshape(rpn_scores, [batch_size, -1, 1])
+            return rpn_boxes, rpn_scores
+
+        if not isinstance(feature_map, (dict, list, tuple)):
+            return call_single_level(feature_map)
+        elif isinstance(feature_map, (list, tuple)):
+            rpn_boxes = []
+            rpn_scores = []
+            for f_map in feature_map:
+                rpn_box, rpn_score = call_single_level(f_map)
+                rpn_boxes.append(rpn_box)
+                rpn_scores.append(rpn_score)
+            rpn_boxes = tf.concat(rpn_boxes, axis=1)
+            rpn_scores = tf.concat(rpn_scores, axis=1)
+            return rpn_boxes, rpn_scores
+        else:
+            rpn_boxes = {}
+            rpn_scores = {}
+            for lvl, f_map in feature_map.items():
+                rpn_box, rpn_score = call_single_level(f_map)
+                rpn_boxes[lvl] = rpn_box
+                rpn_scores[lvl] = rpn_score
+            rpn_boxes = tf.concat(tf.nest.flatten(rpn_boxes), axis=1)
+            rpn_scores = tf.concat(tf.nest.flatten(rpn_scores), axis=1)
+            return rpn_boxes, rpn_scores
+
+    def get_config(self):
+        config = {
+            "num_anchors_per_location": self.num_anchors,
+        }
+        base_config = super().get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+
+# class agnostic regression
+class RCNNHead(tf.keras.layers.Layer):
```

Comment:
lets try to add these to separate files all across the board; it will make your FasterRCNN file easier to read

Reply:
For now they're kept in the same file as a demonstration of "how many LOC is required to build RCNN" model -- but once our model starts to support user passed in RCNNHead, then yeah we should split them.